### PR TITLE
WIP: issue 11 code coverage

### PR DESCRIPTION
**Pull request by kazekagegaara**
This is for issue #11 .
Started working on `set.js` file. The test passes, but code coverage still shows that `if (error) {
          return reject(error)
        }` (line 22 in set.js) isn't covered.

### CODE REVIEW

File: lib/set.js [hunk 1]
```diff
@@ -4,17 +4,18 @@ var format = require('util').format
 
 var Promise = require('lie')
 
-var generateSalt = require('./utils/generate-salt')
-var hashPassword = require('./utils/hash-password')
+var internals = module.exports.internals = {}
+internals.hashPassword = require('./utils/hash-password')
+internals.generateSalt = require('./utils/generate-salt')
```
Comment (gr2m) [inline lib/set.js:9]: This is what I meant with exposing `internals`. This way you you do

``` js
var set = require('./path/to/set')
```

then set.internals.hashPassword is accessible. That way we can mock its behavior in tests

File: tests/unit/set-test.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  set(state, username, password)
+
+  .then(function () {
+    t.error('Should reject')
+  })
+
+  .catch(function (error) {
+    t.is(error.message, 'hashPassword error', 'rejects with error from hashPassword')
+  })
+})
```
Comment (gr2m) [inline tests/unit/set-test.js:54]: when we do unit tests, then we try to only test one method at a time, in this case the `set` method. What you will see a lot in Hoodie’s code is a first argument `state`, which is passed in when initialising the API in the constructor, in this case here: https://github.com/hoodiehq/pouchdb-admins/blob/master/lib/api.js#L16

`set.bind(null, state)` will implicitly pass in the `state` argument each time someone calls `admins.set(username, password)`. So `state` is not accessible from the outside. It’s a style of functional programming. The benefit is that for testing, we can put in what ever `state` we want for the respective test case, which is very helpful

Check out unit tests of [hoodie-account-client](https://github.com/hoodiehq/hoodie-account-client/tree/master/test/unit) to see how we do unit tests.

**Comment by gr2m** [comment]
Hey Manit, I pushed a few updates because I thought it would be the best way to explain. I added a few notes to the changes, I hope everything makes sense? Please let me know if you have any further questions

**Comment by gr2m** [comment]
Hey @kazekagegaara, do you still want to finish this up?

**Comment by minrwhite** [comment]
I think this is failing because nvm 'stable' version has jumped to v7.0.0 and npm packages aren't cleaned out entirely between builds. I'll open an issue.

**Comment by gr2m** [comment]
@benwhite-deltas can you rebase on master? Is this ready for review or still WIP?

**Comment by minrwhite** [comment]
Yeah, I'm happy to do that, when I've got 5 mins. WRT WIP, it depends - are we aiming for 100% with this PR?

**Comment by gr2m** [comment]
we aim for 100% code coverage, but it doesn’t need to be all in this PR

**Comment by minrwhite** [comment]
Rebased, ready for review/merge.

File: tests/unit/validate-password-test.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+var simple = require('simple-mock')
+var test = require('tape')
+
+var rewire = require('rewire')
```
Comment (gr2m) [inline tests/unit/validate-password-test.js]: @benwhite-deltas never used [rewire](https://www.npmjs.com/package/rewire), looks good! We use [proxyquire](https://www.npmjs.com/package/proxyquire) in other projects. Do you see a benefit of one over the other? If not I’d stick with `proxyquire` to make it simpler for contributors to jump between repositories

**Review by gr2m — CHANGES_REQUESTED** [review]

File: tests/unit/validate-password-test.js (hunk 3 above)
Comment (minrwhite) [reply tests/unit/validate-password-test.js]: As all things tend to be, it's a trade off.

I see `rewire` as being better when you need access to view/alter non-exported global vars defined in the module, whereas `proxyquire` allows you to handle modules that are fetched using `require()` - these are possibly intersecting but not identical sets of parameters. I also particularly like `rewire`'s `__with__` for testing purposes.

Looking at how the pull request started, I was attempting to do the kind of testing as done in set-test.js without having to expose an 'internals' object. But given that all of the objects included in that internals object are modules anyway, I agree that reducing cognitive load switching between different workflows, and refactoring to using `proxyquire` exclusively would be best.

Actioning now!

**Review by gr2m — APPROVED** [review]
sweet, thanks 👍 getting there :) [Coverage Status](https://coveralls.io/r/hoodiehq/pouchdb-admins?branch=master)

**Pull request merged**